File: frontend/src/services/api/rates.ts
import { apiClient } from './client'

/**
 * Matches the JSON shape served by the backend at /api/rates/instruments.
 * Source of truth: backend/src/ledger_sync/config/instrument_rates.json.
 */
export interface InstrumentRates {
  updated_at: string
  epf: {
    rate_pct: number
    effective_from: string
    effective_until: string | null
    source_url: string
    notes?: string
  }
  ppf: {
    rate_pct: number
    effective_from: string
    effective_until: string | null
    source_url: string
    notes?: string
  }
  nps: {
    default_allocation_pct: { equity: number; corp_bond: number; govt_bond: number }
    historical_return_pct: { equity: number; corp_bond: number; govt_bond: number }
    effective_from: string
    source_url: string
    notes?: string
  }
}

export const ratesService = {
  async getInstrumentRates(): Promise<InstrumentRates> {
    const res = await apiClient.get<InstrumentRates>('/api/rates/instruments')
    return res.data
  },
}

File: backend/src/ledger_sync/api/rates.py
"""Instrument rates endpoint.

Serves EPF/PPF/NPS rates from a dated JSON config. No reliable public
JSON API exists for these rates, so the file at
``config/instrument_rates.json`` is the source of truth. Update it when
a new rate is notified (EPFO yearly, Ministry of Finance quarterly).
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException

from ledger_sync.api.deps import CurrentUser

router = APIRouter(prefix="/api/rates", tags=["rates"])

_CONFIG_PATH = Path(__file__).resolve().parent.parent / "config" / "instrument_rates.json"


@lru_cache(maxsize=1)
def _load_rates() -> dict[str, Any]:
    if not _CONFIG_PATH.exists():
        raise FileNotFoundError(f"Missing instrument rates config at {_CONFIG_PATH}")
    with _CONFIG_PATH.open(encoding="utf-8") as f:
        data: dict[str, Any] = json.load(f)
    return data


@router.get(
    "/instruments",
    responses={503: {"description": "Instrument rates config unavailable"}},
)
def get_instrument_rates(_current_user: CurrentUser) -> dict[str, Any]:
    """Return EPF/PPF/NPS rates with effective_from and source_url metadata."""
    try:
        return _load_rates()
    except (OSError, json.JSONDecodeError) as e:
        raise HTTPException(
            status_code=503,
            detail="Instrument rates config unavailable",
        ) from e
